Site Editor: Add template switcher to navigation panel

david-szabo97 opened the pull request:
Related: #25262

## Notes about code
- `template-switcher` is a copy of the original `template-switcher` found in the header. 
- `template-switcher/index.js` is overhauled to work with the `NavigationPanel`
- ~Changes in `template-switcher/template-preview.js` and `template-switcher/theme-preview.js` are minimal. Only changed the container's (very-first-div) `className`. **Feel free to ignore these during review**~
- reused as much as possible from the original `template-switcher`
- ~original styles are used and added some overrides in the `navigation-panel/styles.scss`~

## Description
This change replicates the template switcher from the header in the Navigation sidebar.

## How has this been tested?
* `yarn wp-env start`
* Enable FSE
* Open Site Editor
* Click site logo to toggle the navigation sidebar

## Screenshots
![](https://user-images.githubusercontent.com/2256104/94148156-8c337500-fe76-11ea-9ccb-df0545506256.gif)

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

david-szabo97 committed:
Add template switcher to navigation panel

david-szabo97 committed:
Fix theme preview text color

david-szabo97 committed:
Fix position of icons

david-szabo97 committed:
Replace checkmark icon with active state

david-szabo97 commented:
1. Should we remove the original template switcher from the header in this PR? I'm happy to do it but I wasn't sure how far we should go in one PR. Happy to do it and refactor all the styling and code.
2. Since Template switcher is not a nested menu right now, we can't have multiple active items in the root menu

david-szabo97 committed:
Use original component since we don't have an icon anymore

Copons commented on packages/edit-site/src/components/left-sidebar/navigation-panel/template-switcher/index.js:
```diff
@@ -0,0 +1,213 @@
+			) }
+		</>
+	);
+}
```
This looks very similar to `NavigationItemWithIcon`, could we "merge" the 2 components?

david-szabo97 replied:
Kinda different 🤔 I unified them but I'm not sure if it's any better this way. Both way seems a little bit confusing. Maybe it's too smart now.

Copons commented on packages/edit-site/src/components/left-sidebar/navigation-panel/template-switcher/index.js:
```diff
@@ -0,0 +1,213 @@
+		<NavigationItem
+			title={
+				<>
+					{ title }{ ' ' }
```
Is the empty space `{ ' ' }` necessary?

david-szabo97 replied:
I don't think so. Removed it.

Copons commented on packages/edit-site/src/components/left-sidebar/navigation-panel/template-switcher/index.js:
```diff
@@ -0,0 +1,213 @@
+					{ icon && (
+						<div className="edit-site-template-switcher__label-home-icon">
+							<Icon icon={ icon } />
+						</div>
```
`TemplateLabel` wrapped the icon in `Tooltip`.
I think it's a nice and easy a11y win if we brought the same feature here.
We could add an `iconLabel` prop for the tooltip, that would default to `title` if not provided.

david-szabo97 replied:
Done!

Copons commented on packages/edit-site/src/components/left-sidebar/navigation-panel/template-switcher/index.js:
```diff
@@ -0,0 +1,213 @@
+	};
+	const onMouseLeaveTheme = () => {
+		setThemePreviewVisible( () => false );
+	};
```
What about refactoring this a little bit?

Now we don't have a `onHover` anymore, so we should probably define both on enter/on leave for the template part as well.
Also we could simplify `setHoveredTemplate` since it's only ever used for template parts.
Finally, I'm not sure why `setThemePreviewVisible` uses a callback, when it doesn't derive the new value from the previous. 🤔 

E.g.
```jsx
const [ templatePartPreviewVisible, setTemplatePartPreviewVisible ] = useState();
const [ themePreviewVisible, setThemePreviewVisible ] = useState( false );

const onMouseEnterTemplatePart = ( id ) => setTemplatePartPreviewVisible( id );
const onMouseLeaveTemplatePart = ( id ) => setTemplatePartPreviewVisible( null );

const onMouseEnterTheme = ( id ) => themePreviewVisible( true );
const onMouseLeaveTheme = ( id ) => themePreviewVisible( false );
```

Copons commented on packages/edit-site/src/components/left-sidebar/navigation-panel/template-switcher/template-preview.js:
```diff
@@ -0,0 +1,33 @@
+	);
+	const blocks = useMemo(
+		() => ( template ? parse( template?.content?.raw || '' ) : [] ),
+		[ template ]
```
`template` is an object, are we sure this `useMemo` is only updating when `template` changes instead of every render?
Same goes for the `item` dependency of the `useSelect` above.

Do you think it would make more sense to destructure the item ID and use that as dependency?
e.g.
```jsx
function Template Preview( { item } ) {
  const { id: itemId } = item;

  const template = useSelect( () => {}, [ itemId ] );
  const blocks = useMemo( () => {}, [ itemId ] );
}

david-szabo97 replied:
How about using `template?.content?.raw || ''` as a dependency? Probably `itemId` is safe too.

Copons commented on packages/edit-site/src/components/left-sidebar/navigation-panel/template-switcher/template-preview.js:
```diff
@@ -0,0 +1,33 @@
+		<div className="edit-site-navigation-panel__preview">
+			{ !! blocks && (
+				<BlockPreview blocks={ blocks } viewportWidth={ 1200 } />
+			) }
```
What happens if there are no blocks to preview?
What about not rendering the component at all?

Copons commented on packages/edit-site/src/components/left-sidebar/navigation-panel/template-switcher/theme-preview.js:
```diff
@@ -0,0 +1,51 @@
+			/>
+			<div className="edit-site-template-switcher__theme-preview-description">
+				{ truncate(
+					/* Not using description.rendered here, as we might contain after an opening HTML tag. */
```
We should probably clarify this comment 😄

david-szabo97 replied:
Done!

Copons commented on packages/edit-site/src/components/left-sidebar/navigation-panel/template-switcher/template-preview.js:
```diff
@@ -0,0 +1,33 @@
+	);
+}
+
+export default TemplatePreview;
```
Nitpick: we can default-export functions in one line `export default function TemplatePreview() {}`.

Copons commented on packages/edit-site/src/components/left-sidebar/navigation-panel/template-switcher/theme-preview.js:
```diff
@@ -0,0 +1,51 @@
+	);
+}
+
+export default ThemePreview;
```
Nitpick: we can default-export functions in one line `export default function ThemePreview() {}`.

Copons commented on packages/edit-site/src/components/left-sidebar/navigation-panel/template-switcher/index.js:
```diff
@@ -0,0 +1,213 @@
+		label: <TemplateLabel template={ templatePart } />,
+		value: templatePart.id,
+		slug: templatePart.slug,
+	} ) );
```
We probably don't need this map here, as we can just as easily do it once in the render below.
E.g.
```jsx
{ templateParts?.map( ( templatePart ) => (
	<NavigationItem
		key={ `template-part-${ templatePart.id }` }
		item={ templatePart.slug }
		title={ <TemplateLabel template={ templatePart } /> }
		onClick={ () => onActiveTemplatePartIdChange( templatePart.id ) }
		onMouseEnter={ () => onHoverTemplatePart( templatePart.id ) }
		onMouseLeave={ () => onHoverTemplatePart( null ) }
	/>
) ) }
```

Note also that while there are no checks, `title` is supposed to be a string.
I guess that if we [unify `TemplateLabel` and `NavigationItemWithIcon`](https://github.com/WordPress/gutenberg/pull/25615/files#r494383927) this inconsistency wouldn't happen anymore.

david-szabo97 replied:
🤔 If we are supposed to use `string` as `title`, then I need to copy-paste the whole part from the `NavigationItem` to get the same styling with a custom icon.

david-szabo97 committed:
Export one liner

david-szabo97 committed:
Don't render preview if blocks is empty

david-szabo97 committed:
Return directly

david-szabo97 committed:
Spacing

david-szabo97 committed:
Remove empty space

david-szabo97 committed:
Fix template preview

david-szabo97 committed:
Fix template preview

david-szabo97 committed:
Remove unnecessary templateParts map

david-szabo97 committed:
Unify NavigationItemWithIcon and TemplateLabel

david-szabo97 committed:
Fix useMemo, use raw content as dependency

david-szabo97 committed:
Clarify theme description truncate

david-szabo97 committed:
I18n

Copons commented:
@david-szabo97 I totally failed to convey what I was thinking for the `NavigationItemWithIcon`.
My idea was to have them render a custom component, e.g.

```jsx
	return (
		<NavigationItem item={ item } title={ title }>
			<Button { ...props }>
				{ title }
				{ icon && (
					<Tooltip text={ iconLabel || title }>
						<div className="edit-site-template-switcher__label-home-icon">
							<Icon icon={ icon } />
						</div>
					</Tooltip>
				) }
			</Button>
		</NavigationItem>
	);
```

Note that `NavigationItem` keeps the `title` (to be searchable), and gain an `item` prop, so that it can be activated.
All other props (e.g. `onClick`) are passed directly to the `Button` inside.

Copons commented:
> 1. Should we remove the original template switcher from the header in this PR? I'm happy to do it but I wasn't sure how far we should go in one PR. Happy to do it and refactor all the styling and code.

Yeah I guess there shouldn't be any harm here in removing the header's `TemplateSwitcher` (let's keep the `PageSwitcher` though!!). 🙂 

> 2. Since Template switcher is not a nested menu right now, we can't have multiple active items in the root menu

🤔 Do we have a case of multiple active items?
AFAICS we can either activate a template or a template part. Am I missing something?

If we need multiple active items, we should have multiple `Navigation` component, with their own states, btw.

david-szabo97 committed:
Use custom children instead of title

david-szabo97 committed:
Make template an active item as well

david-szabo97 committed:
Fix undefined error

david-szabo97 committed:
Fix icon colors

david-szabo97 committed:
Remove template switcher from header

david-szabo97 committed:
Remove original template switcher code

david-szabo97 committed:
Fix error when theme is not loaded yet

david-szabo97 committed:
Move template switcher style file

david-szabo97 committed:
Remove unused styles

david-szabo97 committed:
Rename styles

david-szabo97 committed:
Add loading placeholder for template parts

david-szabo97 commented:
@Copons 

* Removed original template switcher (with the source as well, I'm not sure if we want to keep it, WDYT?)
* Cleaned up CSS
* Used custom children to avoid setting a component for `title` property (I was driven away by the fact we are losing some styling due to skipping the `title` prop, but it looks alright)
* Template is now an actual item and can be active

> 🤔 Do we have a case of multiple active items?
> AFAICS we can either activate a template or a template part. Am I missing something?

I was thinking about how we are going to put the template switcher and the page switcher on the same navigation. I guess they are going to be nested so we won't have multiple active items.

vindl commented:
> Removed original template switcher (with the source as well, I'm not sure if we want to keep it, WDYT?)

No need to keep the source, we can always bring it back from git history if needed.

> I was thinking about how we are going to put the template switcher and the page switcher on the same navigation. I guess they are going to be nested so we won't have multiple active items.

We don't have to worry about pages for now, just templates.

vindl reviewed (commented):
This tests well for me but we'll need to update the failing e2e tests to account for template switcher changes. We also need to display all templates and not just the ones related to current page content selection, but that can also be done in a follow-up PR.

david-szabo97 committed:
Fix e2e

david-szabo97 committed:
Fix e2e

david-szabo97 committed:
Extract openNavigation function

david-szabo97 commented:
Rebased because of #25622. It added a slide-in animation to the navigation which seems to also occur when we are switching between template and template part 🤔
